README.md
# ISBN

A library to work with ISBN written in Elixir.

[![Hex pm](https://img.shields.io/hexpm/v/isbnex.svg?style=flat)](https://www.hex.pm/packages/isbnex)
[![Hex pm](https://img.shields.io/hexpm/l/isbnex)](https://github.com/renanvy/isbnex/blob/master/LICENSE)
[![Elixir CI](https://github.com/renanvy/isbnex/workflows/build/badge.svg?branch=master)](https://github.com/renanvy/isbnex/workflows/build/badge.svg?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/renanvy/isbnex/badge.svg?branch=master)](https://coveralls.io/github/renanvy/isbnex?branch=master)

The International Standard Book Number (ISBN) is a numeric commercial book identifier which is intended to be unique. Publishers purchase ISBNs from an affiliate of the International ISBN Agency.

An ISBN is assigned to each separate edition and variation (except reprintings) of a publication. For example, an e-book, a paperback and a hardcover edition of the same book will each have a different ISBN. The ISBN is ten digits long if assigned before 2007, and thirteen digits long if assigned on or after 1 January 2007. The method of assigning an ISBN is nation-specific and varies between countries, often depending on how large the publishing industry is within a country. Reference [here](https://en.wikipedia.org/wiki/International_Standard_Book_Number)

This library provides a validation that checks if the ISBN is valid and a formatter.
The ISBN has check digit algorithm, you can check the details [here](https://en.wikipedia.org/wiki/International_Standard_Book_Number).

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `isbnex` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:isbnex, "~> 0.2.0"}
  ]
end
```

## Quick Start

You can verify if a ISBN is valid by calling the function `ISBN.valid?/1`. Works with both ISBN-10 and ISBN-13. Allows hyphens and spaces in the string.

```elixir
iex> ISBN.valid?("9971502100")
true

iex> ISBN.valid?("99-9215-810-7")
true

iex> ISBN.valid?("978-0-306-40615-7")
true

iex> ISBN.valid?("978 0 306 40615 7")
true

iex> ISBN.valid?("0-9752298-0-X")
true

iex> ISBN.valid?("978-0-306-40615-6")
false
```

The `ISBN.format/1` returns ISBN wrapped.

```elixir
iex> ISBN.format("9780306406157")
"978-03-0640-615-7"

iex> ISBN.format("9992158107")
"99-9215-810-7"

iex> ISBN.format("111111")
nil
```

## Validation with Changeset

You can use `ISBN.Ecto.Changeset.validate_isbn/3` function to validate a field in your changeset. Example:

```elixir
defmodule MyApp.Libraries.Book do
  @moduledoc """
  The Book schema.
  """

  use Ecto.Schema
  import Ecto.Changeset

  # Import this module
  import ISBN.Ecto.Changeset

  schema "books" do
    field :isbn, :string, null: false
  end

  @doc false
  def changeset(book, attrs) do
    book
    |> cast(attrs, [:isbn])
    |> validate_isbn([:isbn])
  end
end

iex> changeset = MyApp.Libraries.Book.changeset(%MyApp.Libraries.Book{}, %{isbn: "12345567"})
iex> {:error, %Ecto.Changeset{}} = Repo.insert(changeset)
iex> changeset
#Ecto.Changeset<action: :insert, changes: %{isbn: "1234567890"}, errors: [isbn: {"is invalid", []}], data: #MyApp.Libraries.Book<>, valid?: false>
```

## Docs

The docs can be found at [https://hexdocs.pm/isbnex](https://hexdocs.pm/isbnex).

## Contribute

Feel free to contribute to this project. If you have any suggestions or bug reports just open an issue or a PR.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2020-04-19
### Add
- Provide function to validate ISBN field in a changeset

## [0.1.1] - 2020-04-16
### Add
- Added function to remove spaces and dashes
- Improve docs

test/isbn_test.exs
defmodule ISBNTest do
  use ExUnit.Case, async: true
  doctest ISBN
  alias ISBN

  describe "valid?/1 " do
    test "returns truthy when ISBN is valid" do
      # ISBN10
      assert ISBN.valid?("9971502100")
      assert ISBN.valid?("99921-58-10-7")
      assert ISBN.valid?("960 425 059 0")
      assert ISBN.valid?("0-9752298-0-X")
      assert ISBN.valid?("    0-9752298-0-X    ")

      # ISBN13
      assert ISBN.valid?("9780306406157")
      assert ISBN.valid?("9783161484100")
      assert ISBN.valid?("978-0-306-40615-7")
      assert ISBN.valid?("978 0 306 40615 7")
      assert ISBN.valid?("   978 0 306 40615 7   ")
    end

    test "returns falsy when ISBN is invalid" do
      refute ISBN.valid?("99")
      refute ISBN.valid?(9)

      # ISBN10
      refute ISBN.valid?("9971502101")
      refute ISBN.valid?("9971502101234234233")
      refute ISBN.valid?("99921-58-10-6")

      # ISBN13
      refute ISBN.valid?("9780306406156")
      refute ISBN.valid?("9971502100234234233")
      refute ISBN.valid?("978-0-306-40615-6")
    end
  end

  describe "format/1" do
    test "returns ISBN formatted when is valid" do
      # ISBN10
      assert ISBN.format("9992158107") == "99-9215-810-7"
      assert ISBN.format("99-9215-810-7") == "99-9215-810-7"

      # ISBN13
      assert ISBN.format("9780306406157") == "978-03-0640-615-7"
      assert ISBN.format("978-03-0640-615-7") == "978-03-0640-615-7"
    end

    test "returns nil when ISBN is invalid" do
      assert ISBN.format("9") == nil

      # ISBN10
      assert ISBN.format("9992158106") == nil

      # ISBN13
      assert ISBN.format("9780306406156") == nil
    end

    test "returns nil when ISBN is not an string" do
      assert ISBN.format(9_992_158_107) == nil
    end
  end
end


test/ecto/changeset_test.exs
defmodule ISBN.Ecto.ChangesetTest do
  use ExUnit.Case, async: true

  doctest ISBN.Ecto.Changeset

  import ISBN.Ecto.Changeset

  describe "validate_isbn/2" do
    test "returns error with default message when ISBN is invalid" do
      assert [{:isbn, {"is invalid", []}}] ==
               {%{}, %{isbn: :string}}
               |> Ecto.Changeset.cast(%{"isbn" => "1234567890"}, [:isbn])
               |> validate_isbn(:isbn)
               |> Map.get(:errors)
    end

    test "returns error with custom message when ISBN is invalid" do
      assert [{:isbn, {"isbn is invalid", []}}] ==
               {%{}, %{isbn: :string}}
               |> Ecto.Changeset.cast(%{"isbn" => "1234567890"}, [:isbn])
               |> validate_isbn(:isbn, message: "isbn is invalid")
               |> Map.get(:errors)
    end

    test "non adds error to valid isbn field" do
      assert [] ==
               {%{}, %{isbn: :string}}
               |> Ecto.Changeset.cast(%{"isbn" => "978-03-0640-615-7"}, [:isbn])
               |> validate_isbn(:isbn)
               |> Map.get(:errors)
    end
  end
end


mix.exs
defmodule Isbn.MixProject do
  use Mix.Project

  def project do
    [
      app: :isbnex,
      version: "0.2.0",
      elixir: "~> 1.6",
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],
      start_permanent: Mix.env() == :prod,
      description: "A library to work with ISBN written in Elixir.",
      package: package(),
      elixirc_paths: elixirc_paths(Mix.env()),
      name: "ISBN",
      docs: docs(),
      deps: deps(),
      source_url: "https://github.com/renanvy/isbnex"
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc, "~> 0.14", only: :dev, runtime: false},
      {:excoveralls, "~> 0.10", only: :test},
      {:ecto, "~> 3.4", optional: true}
    ]
  end

  defp package do
    [
      name: "isbnex",
      maintainers: ["Renan Carneiro Oliveira"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/renanvy/isbnex"}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  defp docs do
    [
      source_url: "https://github.com/renan/isbnex",
      extras: ["README.md"]
    ]
  end
end


lib/ecto/changeset.ex
if Code.ensure_loaded?(Ecto.Changeset) do
  defmodule ISBN.Ecto.Changeset do
    @moduledoc """
    Provide function to validate ISBN field in a changeset
    """

    @doc """
    Verifies if given `field` in the `changeset` has a valid ISBN.

    ## Examples

        iex> {%{}, %{isbn: :string}}
        ...> |> Ecto.Changeset.cast(%{isbn: "978-03-0640-615-7"}, [:isbn])
        ...> |> ISBN.Ecto.Changeset.validate_isbn(:isbn)
        #Ecto.Changeset<action: nil, changes: %{isbn: "978-03-0640-615-7"}, errors: [], data: %{}, valid?: true>

        iex> {%{}, %{isbn: :string}}
        ...> |> Ecto.Changeset.cast(%{isbn: "1234567890"}, [:isbn])
        ...> |> ISBN.Ecto.Changeset.validate_isbn(:isbn)
        #Ecto.Changeset<action: nil, changes: %{isbn: "1234567890"}, errors: [isbn: {"is invalid", []}], data: %{}, valid?: false>
    """
    def validate_isbn(changeset, field, options \\ []) do
      Ecto.Changeset.validate_change(changeset, field, fn _, value ->
        case ISBN.valid?(value) do
          true ->
            []

          _ ->
            [{field, {options[:message] || "is invalid", []}}]
        end
      end)
    end
  end
end


lib/isbn.ex
defmodule ISBN do
  @moduledoc """
  Documentation for ISBN.
  """

  @isbn13_multipliers [1, 3, 1, 3, 1, 3, 1, 3, 1, 3, 1, 3]
  @isbn10_multipliers [1, 2, 3, 4, 5, 6, 7, 8, 9]

  @doc """
  Checks if the given string is a valid ISBN.

  Works with both ISBN-10 and ISBN-13. Allows hyphens and spaces in the string.

  ## Examples

      iex> ISBN.valid?("9971502100")
      true

      iex> ISBN.valid?("978-03-0640-615-7")
      true

  """

  def valid?(isbn) when is_binary(isbn) do
    last_digit = isbn |> String.trim() |> String.last()

    digits =
      isbn
      |> remove_spaces_and_dashes()
      |> String.slice(0..-2)
      |> String.codepoints()
      |> Enum.map(&String.to_integer/1)

    reveal_verifier(digits) == last_digit
  end

  def valid?(_), do: false

  @doc """
  Returns ISBN formatted.

  ## Examples

      iex> ISBN.format("9992158107")
      "99-9215-810-7"

      iex> ISBN.format("9992158106")
      nil

  """
  def format(isbn) when not is_binary(isbn), do: nil

  def format(isbn) do
    case valid?(isbn) do
      true ->
        isbn
        |> remove_spaces_and_dashes()
        |> String.codepoints()
        |> do_format()

      _ ->
        nil
    end
  end

  defp do_format(digits) when length(digits) == 10 do
    digits
    |> List.insert_at(2, "-")
    |> List.insert_at(7, "-")
    |> List.insert_at(11, "-")
    |> Enum.join()
  end

  defp do_format(digits) when length(digits) == 13 do
    digits
    |> List.insert_at(3, "-")
    |> List.insert_at(6, "-")
    |> List.insert_at(11, "-")
    |> List.insert_at(15, "-")
    |> Enum.join()
  end

  defp do_format(_isbn), do: nil

  defp reveal_verifier(digits) when length(digits) == 9 do
    acc = calculate(digits, @isbn10_multipliers)
    rem = rem(acc, 11)
    if rem == 10, do: "X", else: Integer.to_string(rem)
  end

  defp reveal_verifier(digits) when length(digits) == 12 do
    acc = calculate(digits, @isbn13_multipliers)
    rem = rem(acc, 10)
    if rem == 0, do: "0", else: Integer.to_string(10 - rem)
  end

  defp reveal_verifier(_isbn), do: nil

  defp calculate(digits, multipliers) do
    multipliers
    |> Enum.zip(digits)
    |> Enum.reduce(0, fn {multiplier, digit}, acc ->
      acc + multiplier * digit
    end)
  end

  defp remove_spaces_and_dashes(isbn) do
    isbn
    |> String.trim()
    |> String.replace("-", "")
    |> String.replace(" ", "")
  end
end
